### PR TITLE
Implement defining remote git dependencies in the skaffold configuration.

### PR DESCRIPTION
**Pull request by gsquared94**
Fixes: #5302 

**Description**

Implements defining remote git repositories as skaffold config dependencies.

```yaml
apiVersion: skaffold/v2beta12
kind: Config
requires:
- git:
    repo: https://github.com/GoogleContainerTools/skaffold
    path: examples/multi-config-microservices/leeroy-app
- git:
    repo: https://github.com/GoogleContainerTools/skaffold
    path: examples/multi-config-microservices/leeroy-web
```

The environment variable `SKAFFOLD_REMOTE_CACHE_DIR` or flag `--remote-cache-dir` specifies the download location for all remote repos. If undefined then it defaults to `~/.skaffold/repos`. 
The repo root directory name is a hash of the repo `uri` and the `branch/ref`.
Every execution of a remote module resets the cached repo to the referenced ref. The default ref is `master`. If `master` is not defined then it defaults to `main`.
The remote config gets treated like a local config after substituting the path with the actual path in the cache directory.
 
Look at the new [`remote-multi-config-microservices`](https://github.com/gsquared94/skaffold/tree/remote-config-dependencies-2/integration/examples/remote-multi-config-microservices) example

### CODE REVIEW

File: integration/examples/remote-multi-config-microservices/README.md
Comment (MarlonGamez) [inline integration/examples/remote-multi-config-microservices/README.md:25]: the `leeroy-web` service doesn't actually have a service resource defined for it, so this part fails. Not sure if we want to add that in this PR or a separate one.

I tried adding a service to this and it seems like the url output by this minikube command doesn't work. Not sure if that's just an issue on my end tho

**Review by MarlonGamez — CHANGES_REQUESTED** [review]
Code and testing is looking really good. Have some comments on the example readme

File: integration/examples/remote-multi-config-microservices/README.md
Comment (gsquared94) [reply integration/examples/remote-multi-config-microservices/README.md]: wow, I checked the history on the `microservices` example and `curl $(minikube service leeroy-web --url)` has been on it since the beginning. However, you're right, this command doesn't work. Maybe the behavior changed on `minikube` sometime in the past. Anyways I've modified this example. Will send a separate PR to fix for the other examples.

**Review by MarlonGamez — APPROVED** [review]
LGTM, thanks for fixing the example :) I won't merge yet as I'm not sure if we want to get another pair of eyes on this since it's a larger PR

File: cmd/skaffold/app/cmd/parse_config.go
Comment (nkubala) [inline cmd/skaffold/app/cmd/parse_config.go:173]:
```suggestion
			return nil, fmt.Errorf("caching remote dependency %s: %w", d.GitRepo, err)
```

**Review by nkubala — COMMENTED** [review]
> Every execution of a remote module resets the cached repo to the referenced ref

let's say I have one git module in my project. I run `skaffold dev`, and make a bunch of local changes to the cloned code which I'm seeing be redeployed to my cluster. I take a break by hitting `ctrl+c`, and then later start up a new `skaffold dev` session - all of the changes I made in my previous session are overwritten because my local cache got reset to the remote ref!

I think this should be configurable. it probably makes sense for this to be reset in some cases, but in others users might be experimenting with some changes to a dependency that they later want to commit, or at least persist. the default behavior here might even be to NOT reset the local cache to the remote ref.

**Comment by gsquared94** [comment]
> > Every execution of a remote module resets the cached repo to the referenced ref
> 
> let's say I have one git module in my project. I run `skaffold dev`, and make a bunch of local changes to the cloned code which I'm seeing be redeployed to my cluster. I take a break by hitting `ctrl+c`, and then later start up a new `skaffold dev` session - all of the changes I made in my previous session are overwritten because my local cache got reset to the remote ref!

To decide on this behavior I think we need to decide what role a remote dependency serves. IMO it's a central store of Skaffold managed repositories. There can be multiple projects sharing the same repository akin to how a go module gets downloaded and shared by projects using them. IMO if I intend to make modification to a code and keep them I'd rather not have them in that managed store and clone them explicitly myself.

If I selectively allow reset, then I can't have multiple skaffold projects sharing this cache. WDYT?

**Comment by gsquared94** [comment]
> > > Every execution of a remote module resets the cached repo to the referenced ref
> > 
> > 
> > let's say I have one git module in my project. I run `skaffold dev`, and make a bunch of local changes to the cloned code which I'm seeing be redeployed to my cluster. I take a break by hitting `ctrl+c`, and then later start up a new `skaffold dev` session - all of the changes I made in my previous session are overwritten because my local cache got reset to the remote ref!
> 
> To decide on this behavior I think we need to decide what role a remote dependency serves. IMO it's a central store of Skaffold managed repositories. There can be multiple projects sharing the same repository akin to how a go module gets downloaded and shared by projects using them. IMO if I intend to make modification to a code and keep them I'd rather not have them in that managed store and clone them explicitly myself.
> 
> If I selectively allow reset, then I can't have multiple skaffold projects sharing this cache. WDYT?

cc. @tejal29, @briandealwis, @marlon-gamez what do you guys think about this?

**Comment by nkubala** [comment]
> IMO if I intend to make modification to a code and keep them I'd rather not have them in that managed store and clone them explicitly myself

one of my thoughts around this was that this feature would be a good way to "trial" out a service in your application, since it's really easy to just plug-and-play any service with a skaffold.yaml. in that case we might want to allow users to make some local modifications without stomping on their changes - but the more i think about this, the more i think this is an uncommon use case.

my opinion is that we should give the ability to users to turn this auto-syncing off with a config field, but we should definitely default to resetting their local changes on every run. you make a great point that these local clones could be shared across projects, in which case local changes could cause unexpected behavior.

**Comment by MarlonGamez** [comment]
I'm in agreement with what Gaurav is saying, and I also agree with Nick that this should be configurable as well.

File: pkg/skaffold/schema/latest/config.go
Comment (briandealwis) [inline pkg/skaffold/schema/latest/config.go:96]: Can we improve this:
```suggestion
	// Names includes specific named configs within the file path. If empty, then all configs in the file are included.
	Names []string `yaml:"configs,omitempty"`
```

File: pkg/skaffold/git/gitutil_test.go
Comment (briandealwis) [inline pkg/skaffold/git/gitutil_test.go]: maps are not sorted, so these could happen in any order.

See [the helm tests](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/deploy/helm/helm_test.go#L466) for an example of how to chain examples.

File: cmd/skaffold/app/cmd/parse_config.go
Comment (briandealwis) [inline cmd/skaffold/app/cmd/parse_config.go:209]: This case shouldn't be possible, right?  So it indicates a logical flaw.  Since errors from here don't seem to be recoverable, a `logrus.Fatal()` seems more appropriate.

**Comment by briandealwis** [comment]
I agree with @gsquared94: the local clones in `~/.skaffold/repos` are shared and should be treated as read-only otherwise you risk interfering with other builds.

But @nkubala is right too: people need the ability make tweaks and overrides (viz `go.mod`'s `replace`).  Especially when using multi-level dependencies.  Users can do local changes by using a different `--remote-cache-dir` (maybe this should be renamed to `--git-remotes`?).  I wonder if we should include the ability for a top-level config to do go.mod-like replacements to override a dependency.

So I'd advocate changing the strict behaviour to logging warnings on finding unsaved work or unpublished commits.

File: pkg/skaffold/git/gitutil_test.go
Comment (gsquared94) [reply pkg/skaffold/git/gitutil_test.go]: already fixed

File: cmd/skaffold/app/cmd/parse_config.go
Comment (gsquared94) [reply cmd/skaffold/app/cmd/parse_config.go:201]: this is just the key to a `map`. Not sure what it has to do with the platform. It's not actually a directory name or anything.

File: cmd/skaffold/app/cmd/parse_config.go
Comment (gsquared94) [reply cmd/skaffold/app/cmd/parse_config.go:52]: done

File: cmd/skaffold/app/cmd/parse_config.go
Comment (gsquared94) [reply cmd/skaffold/app/cmd/parse_config.go]: done

File: pkg/skaffold/git/gitutil.go
Comment (gsquared94) [reply pkg/skaffold/git/gitutil.go]: copied original copyright.

File: pkg/skaffold/git/gitutil.go
Comment (gsquared94) [reply pkg/skaffold/git/gitutil.go]: done

File: cmd/skaffold/app/cmd/parse_config.go
Comment (gsquared94) [reply cmd/skaffold/app/cmd/parse_config.go]: done anyways

File: cmd/skaffold/app/cmd/parse_config.go
Comment (gsquared94) [reply cmd/skaffold/app/cmd/parse_config.go]: done

File: pkg/skaffold/schema/latest/config.go
Comment (gsquared94) [reply pkg/skaffold/schema/latest/config.go]: done

File: cmd/skaffold/app/cmd/parse_config.go
Comment (gsquared94) [reply cmd/skaffold/app/cmd/parse_config.go:62]: I don't think there's any difference either way.

**Pull request merged**
